mlflow/genai/judges/adapters/litellm_adapter.py

```diff
@@ -237,17 +277,36 @@ def _prune_messages_for_context_window():
         try:
             try:
                 response = _invoke_litellm(
-                    litellm_model_uri=litellm_model_uri,
+                    model=model,
                     messages=messages,
                     tools=tools,
                     num_retries=num_retries,
                     response_format=response_format,
                     include_response_format=include_response_format,
                     inference_params=inference_params,
+                    api_base=api_base,
+                    api_key=api_key,
                 )
             except (litellm.BadRequestError, litellm.UnsupportedParamsError) as e:
-                if isinstance(e, litellm.ContextWindowExceededError) or "context length" in str(e):
-                    messages = _prune_messages_for_context_window()
+                error_str = str(e).lower()
+                is_context_window_error = (
+                    isinstance(e, litellm.ContextWindowExceededError)
+                    or "context length" in error_str
+                    or "too many tokens" in error_str
+                )
+                if is_context_window_error:
+                    if provider == "gateway":
+                        # For gateway provider, we don't know the underlying model, so we use
+                        # DSPy-style reactive truncation: remove oldest tool call and retry.
+                        messages = _remove_oldest_tool_call_pair(messages)
+                        if messages is None:
+                            raise MlflowException(
+                                "Context window exceeded and there are no tool calls to truncate. "
+                                "The initial prompt may be too long for the model's context window."
+                            ) from e
+                    else:
+                        # For direct providers, use token-counting based pruning
+                        messages = _prune_messages_for_context_window()
```

Comment:
Can we instead update `_prune_messages_exceeding_context_window_length` so that it can optionally accept the context window length? I'd like to avoid maintaining separate sets of truncation logic.